pennylane/devices/qubit/apply_operation.py

```diff
@@ -231,9 +232,25 @@ def _(op: type_op, state):
     return _apply_operation_default(op, state, is_state_batched, debugger)
 
 
+def apply_operation_csr_matrix(op, state, is_state_batched: bool = False):
+    """The csr_matrix specialized version apply operation."""
+    # State is numpy array, should have been stored in tensor version
+    # remember the initial shape and recover in the end
+    if sp.sparse.issparse(state):
+        raise TypeError("State should not be sparse in default qubit pipeline")
+    original_shape = math.shape(state)
+    num_wires = len(original_shape) - int(is_state_batched)
+    full_state = math.reshape(state, [-1, 2**num_wires])  # expected: [batch_size, 2**num_wires]
+    state_opT = full_state @ op.sparse_matrix(wire_order=range(num_wires)).T
+    state_reshaped = math.reshape(state_opT, original_shape)
```

Comment:
This is great for now, but we may want to also have a backlog task to investigate improving performance by natively applying the matrix without having to expand it out.

Reply:
Yeah very good catch. I actually already had some idea, basically just reshape statevec into matrix, then we do `dot` to get the new one and finally reshape back. Will backlog immediately.

Reply:
sc-85352